### PR TITLE
20457 - SBC-PAY Common Queue

### PR DESCRIPTION
**Pull request by Jxio**
*Issue #:*
https://github.com/bcgov/entity/issues/20457

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

File: pay-api/src/pay_api/services/payment_account.py [hunk 1]
```diff
@@ -818,13 +818,14 @@ def publish_account_mailer_event_on_creation(self):
...
-            gcp_queue_publisher.publish_to_queue(
+            gcp_queue_message.publish_to_queue(
                 QueueMessage(
                     source=QueueSources.PAY_API.value,
                     message_type=message_type,
                     payload=payload,
                     topic=current_app.config.get('ACCOUNT_MAILER_TOPIC')
-                )
+                ),
+                current_app
```
Comment (seeker25) [inline pay-api/src/pay_api/services/payment_account.py:828]: no point in putting in current_app here as a param, you can get it from flask import

File: pay-api/src/pay_api/services/base_payment_system.py [hunk 2]
```diff
@@ -31,12 +31,12 @@
 from pay_api.models import Receipt as ReceiptModel
 from pay_api.models.refunds_partial import RefundPartialLine
 from pay_api.services.cfs_service import CFSService
-from pay_api.services import gcp_queue_publisher
+from pay_api.services.gcp_queue import gcp_queue_message
```
Comment (seeker25) [inline pay-api/src/pay_api/services/base_payment_system.py:34]: gcp_queue_message? reminds me of a datastructure and not a service

File: pay-api/src/pay_api/services/gcp_queue/gcp_queue.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
+# Copyright © 2023 Province of British Columbia
```
Comment (seeker25) [inline pay-api/src/pay_api/services/gcp_queue/gcp_queue.py:1]: Ideally we need a note stating this is going to be replaced, if we're loading it from LEAR's library?

File: pay-api/tests/unit/services/test_gcp_queue.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+# Sample data for testing
+SAMPLE_QUEUE_MESSAGE = QueueMessage(
+    source='test-source',
+    message_type='test-message-type',
+    payload={'key': 'value'},
+    topic='projects/project-id/topics/topic'
+)
+
+
+@pytest.fixture
```
Comment (seeker25) [inline pay-api/tests/unit/services/test_gcp_queue.py:34]: can use autorun argument if needed

File: pay-api/src/pay_api/services/payment_account.py (hunk 1 above)
Comment (seeker25) [reply pay-api/src/pay_api/services/payment_account.py:828]: `from flask import Flask, current_app`

File: pay-queue/src/pay_queue/external/pubsub.py [hunk 5]
```diff
@@ -78,11 +78,11 @@ def init_app(self, app: Flask):
         self.gcp_auth_key = app.config.get('GCP_AUTH_KEY')
         if self.gcp_auth_key:
             try:
-                audience = app.config.get(
+                audience = current_app.config.get(
```
Comment (seeker25) [inline pay-queue/src/pay_queue/external/pubsub.py:81]: What's the difference here between current_app and app?

File: pay-api/src/pay_api/services/gcp_queue/gcp_queue_message.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+class QueueMessage:
+    """Queue message data class."""
+
+    source: str
+    message_type: str
+    payload: dict
+    topic: str
+
+
+def publish_to_queue(queue_message: QueueMessage, app):
```
[the remark this answers is not in the document]
Comment (Jxio) [reply pay-api/src/pay_api/services/gcp_queue/gcp_queue_message.py]: yup I removed it

File: pay-queue/src/pay_queue/external/pubsub.py (hunk 5 above)
Comment (Jxio) [reply pay-queue/src/pay_queue/external/pubsub.py]: I removed this change. filer used current_app, but it changed back when I check today.

File: pay-api/tests/unit/services/test_gcp_queue.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+@pytest.fixture(autouse=True)
+def setup_gcp_queue(mock_publisher_client):
+    """Mock gcp queue setup."""
+    with patch.object(GcpQueue, 'publisher', new_callable=PropertyMock) as mock_publisher:
+        mock_publisher.return_value = mock_publisher_client.return_value
+        yield
+
+
+def test_publish_to_queue_success(mock_publisher_client):
```
Comment (seeker25) [inline pay-api/tests/unit/services/test_gcp_queue.py:56]: autouse=True, I don't think you need this param anymore

File: pay-api/tests/unit/services/test_gcp_queue.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    future_mock.result.return_value = 'message_id'
+    mock_publisher_client.return_value.publish.return_value = future_mock
+
+    # Call the function under test
+    publish_to_queue(SAMPLE_QUEUE_MESSAGE)
+
+    mock_publisher_client.return_value.publish.assert_called_once()
+
+
+def test_publish_to_queue_no_topic(mock_publisher_client, app):
```
Comment (seeker25) [inline pay-api/tests/unit/services/test_gcp_queue.py:69]: autouse=True, I don't think you need this param anymore

**Comment by Jxio** [comment]
local pay-queue unit test:
image
"tests/integration/test_eft_reconciliation.py::test_eft_tdi17_basic_process - assert 4 == 5" exists locally before this PR

**Review by seeker25 — APPROVED** [review]

**Pull request merged**